### PR TITLE
Enhancements to the adaptive importance sampler

### LINKED ISSUE #21109: Enhancements to the adaptive importance sampler
## Reason
The following enhancements to the adaptive importance sampler classes are made:
- [x] Fix inputs issue in the AdaptiveImportanceSampler class. Currently, the inputs are associated with wrong outputs of the subapp.
- [x] Add a new reporter class to compute statistics of the adaptive importance sampling like the failure probability, coefficient of variation, and moments of the importance distribution.
- [x] Create tests and docs
- [] Verify implementation for an established problem.  

## Design
Improvements are made to the AdaptiveImportanceSampler class. A new AdaptiveImportanceStats Reporter class is created.

## Impact
AdaptiveImportanceSampler accuracy is improved.

### PR DESCRIPTION
**Pull request by somu15**
## Reason
The following enhancements to the adaptive importance sampler classes are made:
- [x] Fix inputs issue in the AdaptiveImportanceSampler class. Currently, the inputs are associated with wrong outputs of the subapp.
- [x] Add a new reporter class to compute statistics of the adaptive importance sampling like the failure probability, coefficient of variation, and moments of the importance distribution.
- [x] Create tests and docs
- [x] Verify implementation for an established problem  

## Design
Improvements are made to the AdaptiveImportanceSampler class. A new AdaptiveImportanceStats Reporter class is created.

## Impact
AdaptiveImportanceSampler accuracy is improved.

closes #21109

### CODE REVIEW

**Comment by somu15** [comment]
Also tagging @yfche

File: modules/stochastic_tools/doc/content/source/reporters/AdaptiveImportanceStats.md [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+- +Coefficient of variation (cov)+
+
+  This quantity is computed as the ratio of square root of variance of pf to the
+  mean pf. The variance of pf is given by [!cite](au1999new):
+
+  \begin{equation}
+    \label{eqn:ais_stats2}
+    \textrm{Var}(\hat{P}_f^{\textrm{AIS}}) = \frac{1}{N}~\Bigg\{\frac{1}{N}~\sum_{i=1}^{N}\Bigg[I\big(F_i(\pmb{x})\geq \mathcal{F}\big)~\frac{q(\pmb{x})}{f(\pmb{x})}\Bigg]^2-\big(\hat{P}_f^{\textrm{AIS}}\big)^2\Bigg\}
+  \end{equation}
```
Comment (zachmprince) [inline modules/stochastic_tools/doc/content/source/reporters/AdaptiveImportanceStats.md:35]:
```suggestion
  !equation id=eqn:ais_stats2
  \textrm{Var}(\hat{P}_f^{\textrm{AIS}}) = \frac{1}{N}~\Bigg\{\frac{1}{N}~\sum_{i=1}^{N}\Bigg[I\big(F_i(\pmb{x})\geq \mathcal{F}\big)~\frac{q(\pmb{x})}{f(\pmb{x})}\Bigg]^2-\big(\hat{P}_f^{\textrm{AIS}}\big)^2\Bigg\}
```

File: modules/stochastic_tools/doc/content/source/reporters/AdaptiveImportanceStats.md [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+  The cov is given by:
+
+  \begin{equation}
+    \label{eqn:ais_stats3}
+    \hat{\delta}^{\textrm{AIS}} = \frac{\sqrt{\textrm{Var}(\hat{P}_f^{\textrm{AIS}})}}{\hat{P}_f^{\textrm{AIS}}}
+  \end{equation}
```
Comment (zachmprince) [inline modules/stochastic_tools/doc/content/source/reporters/AdaptiveImportanceStats.md:42]:
```suggestion
  !equation id=eqn:ais_stats3
  \hat{\delta}^{\textrm{AIS}} = \frac{\sqrt{\textrm{Var}(\hat{P}_f^{\textrm{AIS}})}}{\hat{P}_f^{\textrm{AIS}}}
  \end{equation}
```

File: modules/stochastic_tools/src/samplers/AdaptiveImportanceSampler.C [hunk 3]
```diff
@@ -119,10 +119,34 @@ AdaptiveImportanceSampler::getOutputLimit() const
...
+const Real &
+AdaptiveImportanceSampler::getStdFactor() const
+{
+  return _std_factor;
+}
+
 Real
 AdaptiveImportanceSampler::computeSample(dof_id_type /*row_index*/, dof_id_type col_index)
 {
-  const bool sample = _step > 1 && col_index == 0 && _check_step != _step;
```
Comment (zachmprince) [inline modules/stochastic_tools/src/samplers/AdaptiveImportanceSampler.C:125]: I'm a little confused on why this change does anything.

File: modules/stochastic_tools/test/tests/samplers/AdaptiveImportanceSampler/sub.i [hunk 4]
```diff
@@ -26,13 +26,13 @@
     type = DirichletBC
     variable = u
     boundary = left
-    value = 0
+    value = 0.028780772062869442
```
Comment (zachmprince) [inline modules/stochastic_tools/test/tests/samplers/AdaptiveImportanceSampler/sub.i:29]: Could you make a quick comment on why this is needed?

File: modules/stochastic_tools/test/tests/samplers/AdaptiveImportanceSampler/tests [hunk 5]
```diff
@@ -9,4 +9,19 @@
     input = 'ais.i'
     jsondiff = 'ais_out.json'
   []
+  [ais_errors]
+    requirement = 'The system shall throw an error when '
+    [ais_error1]
+      type = RunException
+      input = 'ais_error1.i'
+      detail = 'the selected sampler type is not of an adaptive type. '
```
Comment (zachmprince) [inline modules/stochastic_tools/test/tests/samplers/AdaptiveImportanceSampler/tests:17]:
```suggestion
      detail = 'the selected sampler type is not of an adaptive type when performing adaptive decision; '
```

File: modules/stochastic_tools/test/tests/samplers/AdaptiveImportanceSampler/tests [hunk 6]
```diff
@@ -9,4 +9,19 @@
...
+      expect_err = "The selected sampler is not an adaptive sampler."
+    []
+    [ais_error2]
+      type = RunException
+      input = 'ais_error2.i'
+      detail = 'the selected sampler is not of type AdaptiveImportance. '
```
Comment (zachmprince) [inline modules/stochastic_tools/test/tests/samplers/AdaptiveImportanceSampler/tests:23]:
```suggestion
      detail = 'the selected sampler is not of type AdaptiveImportance when reporting adaptive sampling statistics;'
```

File: modules/stochastic_tools/test/tests/samplers/AdaptiveImportanceSampler/ais_error2.i [hunk 7]
```diff
@@ -0,0 +1,72 @@
+[StochasticTools]
```
Comment (zachmprince) [inline modules/stochastic_tools/test/tests/samplers/AdaptiveImportanceSampler/ais_error2.i:1]: Let's make this input as simple as possible:
```
[StochasticTools]
[]

[Samplers/sample]
  type = CartesianProduct
  linear_space_items = '0 1 1'
[]

[Reporters]
  [constant]
    type = ConstantReporter
    real_vector_names = 'reporter_transfer:average:value'
    real_vector_values = '0'
  []
  [ais_stats]
    type = AdaptiveImportanceStats
    output_value = constant/reporter_transfer:average:value
    sampler = sample
  []
[]
```

File: modules/stochastic_tools/test/tests/samplers/AdaptiveImportanceSampler/ais_error1.i [hunk 8]
```diff
@@ -0,0 +1,73 @@
+[StochasticTools]
```
Comment (zachmprince) [inline modules/stochastic_tools/test/tests/samplers/AdaptiveImportanceSampler/ais_error1.i:1]: Let's make this input as simple as possible:
```
[StochasticTools]
[]

[Samplers/sample]
  type = CartesianProduct
  linear_space_items = '0 1 1'
[]

[Reporters]
  [constant]
    type = ConstantReporter
    real_vector_names = 'reporter_transfer:average:value'
    real_vector_values = '0'
  []
  [adaptive_MC]
    type = AdaptiveMonteCarloDecision
    output_value = constant/reporter_transfer:average:value
    inputs = 'inputs'
    sampler = sample
  []
[]
```

File: modules/stochastic_tools/src/reporters/AdaptiveImportanceStats.C [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "AdaptiveImportanceStats.h"
+#include "Sampler.h"
+#include "Distribution.h"
+#include "Normal.h"
+#include "libmesh/utility.h"
+
+registerMooseObject("StochasticToolsApp", AdaptiveImportanceStats);
```
Comment (zachmprince) [inline modules/stochastic_tools/src/reporters/AdaptiveImportanceStats.C:16]: Just a thought. What if we added this in the `AdaptiveMonteCarloDecision` instead of creating new object? `AdaptiveMonteCarloDecision` is already a reporter so we can create these reporter values based on whether the parameters are specified by the user. You can let me know if I'm crazy.

File: modules/stochastic_tools/src/samplers/AdaptiveImportanceSampler.C [hunk 10]
```diff
@@ -119,10 +119,34 @@ AdaptiveImportanceSampler::getOutputLimit() const
...
+AdaptiveImportanceSampler::getDistributionNames() const
+{
+  return _distributions;
+}
+
+const Real &
+AdaptiveImportanceSampler::getStdFactor() const
+{
+  return _std_factor;
+}
```
Comment (zachmprince) [inline modules/stochastic_tools/src/samplers/AdaptiveImportanceSampler.C:144]: I would just put all this in the header file.

File: modules/stochastic_tools/src/reporters/AdaptiveImportanceStats.C [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+AdaptiveImportanceStats::AdaptiveImportanceStats(const InputParameters & parameters)
+  : GeneralReporter(parameters),
+    _output_value(getReporterValue<std::vector<Real>>("output_value", REPORTER_MODE_DISTRIBUTED)),
+    _mu_imp(declareValue<std::vector<Real>>("mu_imp")),
+    _std_imp(declareValue<std::vector<Real>>("std_imp")),
+    _pf(declareValue<std::vector<Real>>("pf")),
+    _cov_pf(declareValue<std::vector<Real>>("cov_pf")),
+    _step(getCheckedPointerParam<FEProblemBase *>("_fe_problem_base")->timeStep()),
+    _sampler(getSampler("sampler")),
+    _ais(dynamic_cast<const AdaptiveImportanceSampler *>(&_sampler)),
```
Comment (zachmprince) [inline modules/stochastic_tools/src/reporters/AdaptiveImportanceStats.C:45]: Is this object going to be able to accept different samplers in the future. If not:
```suggestion
    _sampler(getSampler<AdaptiveImportanceSampler>("sampler")),
```

File: modules/stochastic_tools/src/reporters/AdaptiveImportanceStats.C [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+  _pf_sum = 0.0;
+  _var_sum = 0.0;
+  _distributions_store = _ais->getDistributionNames();
+  _factor = _ais->getStdFactor();
+}
+
+void
+AdaptiveImportanceStats::execute()
+{
+  if (_sampler.getNumberOfLocalRows() == 0 || _check_step == _step)
```
Comment (zachmprince) [inline modules/stochastic_tools/src/reporters/AdaptiveImportanceStats.C:66]: We should probably figure out what will happen in parallel when a processor doesn't own a row.

**Review by zachmprince — CHANGES_REQUESTED** [review]
Initial review. Let's talk in person @somu15

File: modules/stochastic_tools/src/reporters/AdaptiveImportanceStats.C (hunk 11 above)
Comment (somu15) [reply modules/stochastic_tools/src/reporters/AdaptiveImportanceStats.C:45]: In this change, we are not getting `_ais` are we? `_ais` is subsequently used in the code.

File: modules/stochastic_tools/src/reporters/AdaptiveImportanceStats.C (hunk 9 above)
Comment (somu15) [reply modules/stochastic_tools/src/reporters/AdaptiveImportanceStats.C:16]: For now, as per our discussion, I didn't make any changes to have a generalized statistics class. We should have a discussion with the team on how this will work for both regular and adaptive Monte Carlo sampling since the mechanism to compute the statistics are different across different methods. Please let me know what you think.

File: modules/stochastic_tools/src/reporters/AdaptiveImportanceStats.C (hunk 11 above)
Comment (zachmprince) [reply modules/stochastic_tools/src/reporters/AdaptiveImportanceStats.C:45]: With this change you are getting a reference to `AdaptiveImportanceSampler`. You just need to change type of `_sampler` in the header file. This will also throw an error if the sampler is not AIS, so you don't need the error below.

Comment (somu15) [reply modules/stochastic_tools/src/reporters/AdaptiveImportanceStats.C]: I have addressed this now.

**Review by zachmprince — APPROVED** [review]
